src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -0,0 +1,147 @@
+import java.nio.ByteBuffer;
+import java.nio.channels.AsynchronousFileChannel;
+import java.nio.channels.CompletionHandler;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * AsyncCache implementation that uses Java NIO's AsynchronousFileChannel to perform asynchronous
+ * disk reads and writes. This should only be used by devices with an API level of 26 or above.
+ */
+@SuppressLint("NewApi")
+public class DiskBasedAsyncCache extends AsyncCache {
+
+    /** Map of the Key, CacheHeader pairs */
+    private final Map<String, CacheHeader> mEntries = new LinkedHashMap<>(16, .75f, true);
+
+    /** The supplier for the root directory to use for the cache. */
+    private final DiskBasedCacheUtility.FileSupplier mRootDirectorySupplier;
+
+    /** Total amount of space currently used by the cache in bytes. */
+    private long mTotalSize = 0;
+
+    /**
+     * Constructs an instance of the DiskBasedAsyncCache at the specified directory.
+     *
+     * @param rootDirectory The root directory of the cache.
+     */
+    public DiskBasedAsyncCache(final File rootDirectory) {
+        mRootDirectorySupplier =
+                new DiskBasedCacheUtility.FileSupplier() {
+                    @Override
+                    public File get() {
+                        return rootDirectory;
+                    }
+                };
+    }
+
+    /** Clears the cache. Deletes all cached files from disk. */
+    @Override
+    public synchronized void clear() {
+        // TODO (sphill99): Implement
+    }
+
+    /** Returns the cache entry with the specified key if it exists, null otherwise. */
+    @Override
+    public void get(String key, OnGetCompleteCallback callback) {
+        final OnGetCompleteCallback cb = callback;
+        final CacheHeader entry = mEntries.get(key);
+        // if the entry does not exist, return null.
+        if (entry == null) {
+            cb.onGetComplete(null);
+            return;
+        }
+        final File file = getFileForKey(key);
+        final int size = (int) file.length();
+        Path path = Paths.get(file.getPath());
+        AsynchronousFileChannel afc = null;
+        try {
+            afc = AsynchronousFileChannel.open(path, StandardOpenOption.READ);
+            ByteBuffer buffer = ByteBuffer.allocate(size);
+            try {
+                afc.read(
+                        /* destination= */ buffer,
+                        /* position= */ 0,
+                        /* attachment= */ buffer,
+                        new CompletionHandler<Integer, ByteBuffer>() {
+                            @Override
+                            public void completed(Integer result, ByteBuffer attachment) {
+                                // if the file size changes, return null
+                                if (size != file.length()) {
+                                    VolleyLog.d(
```

Comment:
logging stuff:

1) use e instead of d - this is an error, not debugging info
2) You can do `VolleyLog.e("File changed while reading: %s", file.getAbsolutePath())` - no need to use format args for a constant string. Also, I think it's clearer to have the error be the first thing, and specific details like the file name come second.
3) Here and throughout, it should be "%s", not "s%", no?